client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java

```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
         return response.getStatusLine().getStatusCode() == 200;
     }
 
+    private Cancellable performClientRequestAsync(Request request, ResponseListener listener) {
+
+        ListenableFuture<Optional<String>> versionCheck = getVersionValidationFuture();
+
+        CompletableFuture<Void> cancelFlag = new CompletableFuture<Void>();
+        Cancellable result = new Cancellable() {
+            @Override
+            public void cancel() {
+                cancelFlag.complete(null);
+            }
+
+            @Override
+            void runIfNotCancelled(Runnable runnable) {
+                if (cancelFlag.isDone()) {
```

Comment:
The `cancelFlag` future is completed (and not cancelled) when the `Cancellable` returned as a result is cancelled. This is because `CompletableFuture` allows to listen to normal and exceptional completion but not on cancellation. Also there's `isCancelled` and `isCompletedExceptionally` but no `isCompletedSuccessfully`. So `isDone` works here, since the only thing we ever do to this future is to `complete(null)`.

The comments added in e0e0638 provide more details on this flag.

Reply:
Actually the above comment is wrong: cancelling a `CompletableFuture` completes it exceptionnally with a `CancellationException`. I'll fix the comment, but woud like to keep the current behavior as we're not cancelling the flag but raising it.